src/aggregator/client/queue.go

```diff
@@ -204,22 +223,61 @@ func (q *queue) Close() error {
 	return nil
 }
 
-func (q *queue) drain() {
+func (q *queue) drain(tC <-chan (time.Time)) {
 	defer q.wg.Done()
+	defer q.conn.Close()
+	lastDrain := time.Now()
+
+	for {
+		select {
+		case qitem := <-q.bufCh:
+			var needFlush bool
+			msg := qitem.Bytes()
+			q.buf = append(q.buf, msg...)
+			qitem.Close()
+
+			if len(q.buf) > q.batchSize {
+				needFlush = true
+			} else {
+				needFlush = time.Since(lastDrain) >= q.writeInterval
+			}
 
-	for buf := range q.bufCh {
-		if err := q.writeFn(buf.Bytes()); err != nil {
-			q.log.Error("write data error",
-				zap.String("instance", q.instance.Endpoint()),
-				zap.Error(err),
-			)
-			q.metrics.connWriteErrors.Inc(1)
-		} else {
-			q.metrics.connWriteSuccesses.Inc(1)
+			if !needFlush {
+				continue
+			}
+
+			if err := q.writeFn(q.buf); err != nil {
+				q.log.Error("error writing data", zap.String("instance", q.instance.Endpoint()), zap.Error(err))
+				q.metrics.connWriteErrors.Inc(1)
+			} else {
+				q.metrics.connWriteSuccesses.Inc(1)
+			}
```

Comment:
I would extract this method so you can call it once from both this part of the loop and below.

I understand you think it might prevent inlining, but I don't think it's quite enough in the hot path to be an issue.  There's plenty of function calls per metric already in this code path.